### PR TITLE
NMS-15816: Database deadlock triggered by NodeRestService

### PR DESCRIPTION
**Pull request by j-white**
JIRA: [NMS-15816](https://opennms.atlassian.net/browse/NMS-15816)

Avoid acquiring JDBC connection while in the context of a Spring transaction transaction - process one after the other instead.

Also add code that can be used to detect these problems, default to disabled.

[NMS-15816]: https://opennms.atlassian.net/browse/NMS-15816?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by christianpape — APPROVED** [review]
LGTM. It would be nice to have the new property documented somewhere, e.g. opennms.properties or in the troubleshooting section in the docs.

**Comment by j-white** [comment]
> LGTM. It would be nice to have the new property documented somewhere, e.g. opennms.properties or in the troubleshooting section in the docs.

Agreed, I was debating where to put it though. I see it as some advanced flag that we would only ask users to turn on if/when we detect similar symptoms.

File: docs/modules/troubleshooting/pages/other-issues.adoc [hunk 1]
```diff
@@ -86,3 +86,42 @@ The `config-tester` command is even more exhaustive, but it may not catch all er
...
+
+To do so, enable deadlock detection with the following:
+[source, shell]
+----
+# echo 'org.opennms.core.db.deadlock.detection=true' >> /opt/opennms/etc/opennms.properties.d/db.properties
+----
+
+And restart {page-component-title} for the changes to become effective.
+
+Once enabled, you should see the following appear in the logs:
```
Comment (Bonrob2) [inline docs/modules/troubleshooting/pages/other-issues.adoc:102]:
```suggestion
Once enabled, you should see the following in the logs:
```

File: docs/modules/troubleshooting/pages/other-issues.adoc [hunk 2]
```diff
@@ -86,3 +86,42 @@ The `config-tester` command is even more exhaustive, but it may not catch all er
...
+
+If the system ends up in a state where all available database connections are exhausted and fails to recover, you can turn on deadlock detection to help identify the cause.
+
+To do so, enable deadlock detection with the following:
+[source, shell]
+----
+# echo 'org.opennms.core.db.deadlock.detection=true' >> /opt/opennms/etc/opennms.properties.d/db.properties
+----
+
+And restart {page-component-title} for the changes to become effective.
```
Comment (Bonrob2) [inline docs/modules/troubleshooting/pages/other-issues.adoc:100]:
```suggestion
Restart {page-component-title} for the changes to become effective.
```

File: docs/modules/troubleshooting/pages/other-issues.adoc [hunk 3]
```diff
@@ -86,3 +86,42 @@ The `config-tester` command is even more exhaustive, but it may not catch all er
...
+
+Once enabled, you should see the following appear in the logs:
+[source, shell]
+----
+2023-07-03 21:38:49,874 ERROR [Main] o.o.c.d.HikariCPConnectionFactory: Deadlock detection is enabled.
+----
+
+Possible deadlocks will be logged with a message similar to:
```
Comment (Bonrob2) [inline docs/modules/troubleshooting/pages/other-issues.adoc:108]:
```suggestion
Possible deadlocks will be logged with a message similar to the following:
```

File: docs/modules/troubleshooting/pages/other-issues.adoc [hunk 4]
```diff
@@ -86,3 +86,42 @@ The `config-tester` command is even more exhaustive, but it may not catch all er
...
+        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
+        at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
+        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:302) ~[org.apache.servicemix.bundles.spring-aop-4.2.9.RELEASE_1.ONMS.1.jar:?]
+        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202) ~[org.apache.servicemix.bundles.spring-aop-4.2.9.RELEASE_1.ONMS.1.jar:?]
+        at com.sun.proxy.$Proxy136.getNodeIPAddressServiceMap(Unknown Source) ~[?:?]
+        at org.opennms.netmgt.dao.support.DefaultFilterWatcher$FilterSession.lambda$refreshNow$0(DefaultFilterWatcher.java:228) ~[opennms-dao-31.0.3-SNAPSHOT.jar:?]
+        at org.opennms.netmgt.dao.hibernate.DefaultSessionUtils.withManualFlush(DefaultSessionUtils.java:81) ~[opennms-dao-31.0.3-SNAPSHOT.jar:?]
+---
+
+These messages and stack traces should be inspected further to see if there are any potential problems.
```
Comment (Bonrob2) [inline docs/modules/troubleshooting/pages/other-issues.adoc:126]:
```suggestion
Inspect these messages and stack traces further to see if there are any potential problems.
```

**Review by Bonrob2 — CHANGES_REQUESTED** [review]
A few minor changes. Thanks!

**Review by Bonrob2 — APPROVED** [review]
docs LGTM! Thanks!

**Pull request merged**